lib/Storage/Driver/Xml/HistoryIterator.php

```diff
@@ -0,0 +1,192 @@
+                }
+            }
+
+            if ($this->days->valid()) {
+                $this->entries = $this->getEntryIterator();
+            }
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function key()
+    {
+        $this->init();
+        $this->entries->key();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function rewind()
+    {
+        $this->init();
+        $this->entries->rewind();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function valid()
+    {
+        $this->init();
+        return $this->entries->valid();
+    }
+
+    private function init()
+    {
+        if ($this->initialized) {
+            return;
+        }
+
+        $this->initialized = true;
+
+        $this->years = $this->getDirectoryIterator($this->path);
+        $this->months = $this->getDirectoryIterator($this->years->current());
+        $this->days = $this->getDirectoryIterator($this->months->current());
+        $this->entries = $this->getEntryIterator();
+    }
+
+    private function getEntryIterator()
+    {
+        $files = $this->days->current();
+        $files = new \DirectoryIterator($this->days->current());
+        $historyEntries = [];
+        foreach ($files as $file) {
+            if (!$file->isFile()) {
+                continue;
+            }
+
+            if ($file->getExtension() !== 'xml') {
+                continue;
+            }
+
+            $historyEntries[] = $this->getHistoryEntry($file->getPathname());
+        }
+        usort($historyEntries, function ($entry1, $entry2) {
+            if ($entry1->getDate()->format('c') === $entry2->getDate()->format('c')) {
+                return null;
+            }
+
+            return $entry1->getDate()->format('c') < $entry2->getDate()->format('c');
+        });
+        return new \ArrayIterator($historyEntries);
+    }
+
+    private function getHistoryEntry($path)
+    {
+        $dom = new Document();
+        $dom->load($path);
+        $collection = $this->xmlDecoder->decode($dom);
+        $suites = $collection->getSuites();
+        $suite = reset($suites);
+        $envInformations = $suite->getEnvInformations();
+
+        $vcsBranch = null;
+        if (isset($envInformations['vcs']['branch'])) {
+            $vcsBranch = $envInformations['vcs']['branch'];
+        }
+        
+        $summary = $suite->getSummary();
+        $entry = new HistoryEntry(
+            $suite->getUuid(),
+            $suite->getDate(),
+            $suite->getContextName(),
+            $vcsBranch,
+            $summary->getNbSubjects(),
+            $summary->getNbIterations(),
+            $summary->getNbRevolutions(),
+            $summary->getMinTime(),
+            $summary->getMaxTime(),
+            $summary->getMeanTime(),
+            $summary->getMeanRelStDev(),
+            $summary->getTotalTime()
+        );
+
+        return $entry;
```

Comment:
You mean using XPath? Would be much faster, although means duplicating
logic, although maybe the XmlDecoder could provide an API for it. But
yeah, can leave it for now.

On Sun, Mar 20, 2016 at 04:01:49AM -0700, Marco Pivetta wrote:
>    In [1]lib/Storage/Driver/Xml/HistoryIterator.php:
> 
>  > +        $entry = new HistoryEntry(
>  > +            $suite->getUuid(),
>  > +            $suite->getDate(),
>  > +            $suite->getContextName(),
>  > +            $vcsBranch,
>  > +            $summary->getNbSubjects(),
>  > +            $summary->getNbIterations(),
>  > +            $summary->getNbRevolutions(),
>  > +            $summary->getMinTime(),
>  > +            $summary->getMaxTime(),
>  > +            $summary->getMeanTime(),
>  > +            $summary->getMeanRelStDev(),
>  > +            $summary->getTotalTime()
>  > +        );
>  > +
>  > +        return $entry;
> 
>    Instead of decoding and re-building the state, letting xsltproc doing it
>    (much faster at it). Out of scope tho.
> 
>    —
>    You are receiving this because you authored the thread.
>    Reply to this email directly or [2]view it on GitHub
> 
>    Reverse link: [3]unknown
> 
> References
> 
>    Visible links
>    1. https://github.com/phpbench/phpbench/pull/376#discussion_r56762408
>    2. https://github.com/phpbench/phpbench/pull/376/files/18e4fe5aca2ae91e487d82f315100634461e7994#r56762408
>    3. https://github.com/phpbench/phpbench/pull/376/files/18e4fe5aca2ae91e487d82f315100634461e7994#r56762408